File: 30 Days of Code/Day 18/QueuesAndStacks.java
import java.io.*;
import java.util.*;
import java.util.Stack;
import java.util.Queue;

public class QueuesAndStacks {
    // Write your code here.
    // ArrayDeque is faster than Stack 
    ArrayDeque <Character> stack = new ArrayDeque();
    ArrayDeque <Character> queue = new ArrayDeque();

    void pushCharacter(char ch){
        stack.push(ch);
    }
    char popCharacter(){
        return stack.pop();
    }
    void enqueueCharacter(char ch){
        queue.add(ch);
    }
    char dequeueCharacter(){
        return queue.remove();
    }



    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String input = scan.nextLine();
        scan.close();

        // Convert input String to an array of characters:
        char[] s = input.toCharArray();

        // Create a Solution object:
        QueuesAndStacks p = new QueuesAndStacks();

        // Enqueue/Push all chars to their respective data structures:
        for (char c : s) {
            p.pushCharacter(c);
            p.enqueueCharacter(c);
        }

        // Pop/Dequeue the chars at the head of both data structures and compare them:
        boolean isPalindrome = true;
        for (int i = 0; i < s.length/2; i++) {
            if (p.popCharacter() != p.dequeueCharacter()) {
                isPalindrome = false;                
                break;
            }
        }

        //Finally, print whether string s is palindrome or not.
        System.out.println( "The word, " + input + ", is " 
                           + ( (!isPalindrome) ? "not a palindrome." : "a palindrome." ) );
    }
}

File: 30 Days of Code/Day 18/QueuesAndStacks.cpp
// https://www.hackerrank.com/challenges/30-more-exceptions/problem

#include <iostream>
#include <stack>
#include <queue>
using namespace std;
#define MAX 10000

class Solution {
    //Write your code here
    private:
        stack<char> s;
        queue<char> q;
    
    public:
        char popCharacter(){
            char temp=s.top();
            s.pop();
            return temp;
        }

        void pushCharacter(char ch){
            s.push(ch);
        }

        void enqueueCharacter(char ch){
            q.push(ch);
        }

        char dequeueCharacter(){
            char temp=q.front();
            q.pop();
            return temp;
        }
};

int main() {
    // read the string s.
    string s;
    getline(cin, s);
    
  	// create the Solution class object p.
    Solution obj;
    
    // push/enqueue all the characters of string s to stack.
    for (int i = 0; i < s.length(); i++) {
        obj.pushCharacter(s[i]);
        obj.enqueueCharacter(s[i]);
    }
    
    bool isPalindrome = true;
    
    // pop the top character from stack.
    // dequeue the first character from queue.
    // compare both the characters.
    for (int i = 0; i < s.length() / 2; i++) {
        if (obj.popCharacter() != obj.dequeueCharacter()) {
            isPalindrome = false;
            
            break;
        }
    }
    
    // finally print whether string s is palindrome or not.
    if (isPalindrome) {
        cout << "The word, " << s << ", is a palindrome.";
    } else {
        cout << "The word, " << s << ", is not a palindrome.";
    }
    
    return 0;
}